pkg/pushsync/pushsync.go

```diff
@@ -352,129 +365,173 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk, origin bo
 			if errors.Is(err, topology.ErrWantSelf) {
 
 				if !ps.warmedUp() {
-					return nil, ErrWarmup
+					return swarm.ZeroAddress, ErrWarmup
 				}
 
 				if !ps.topologyDriver.IsWithinDepth(ch.Address()) {
-					return nil, ErrOutOfDepthStoring
+					return swarm.ZeroAddress, ErrOutOfDepthStoring
 				}
 
 				ps.pushToNeighbourhood(ctx, fullSkipList, ch, origin, originAddr)
-				return nil, err
+				return swarm.ZeroAddress, err
 			}
-			return nil, fmt.Errorf("closest peer: %w", err)
+
+			return swarm.ZeroAddress, fmt.Errorf("closest peer: %w", err)
 		}
-		ps.metrics.TotalSendAttempts.Inc()
 
-		ctxd, canceld := context.WithTimeout(ctx, defaultTTL)
-		defer canceld()
+		return peer, nil
+	}
 
-		now := time.Now()
-		r, attempted, err := ps.pushPeer(ctxd, peer, ch, origin)
-		var status string
-		if err != nil {
-			status = "failure"
-		} else {
-			status = "success"
-		}
-		ps.metrics.PushToPeerTime.WithLabelValues(status).Observe(time.Since(now).Seconds())
+	for {
+		select {
+		case <-timer.C:
 
-		// attempted is true if we get past accounting and actually attempt
-		// to send the request to the peer. If we dont get past accounting, we
-		// should not count the retry and try with a different peer again
-		if attempted {
-			allowedRetries--
-		}
-		if err != nil {
-			var timeToSkip time.Duration
-			switch {
-			case errors.Is(err, accounting.ErrOverdraft):
-				skipPeers = append(skipPeers, peer)
-			default:
-				timeToSkip = sanctionWait
+			retries--
+
+			peer, err := nextPeer()
+			if err != nil {
+				return nil, err
 			}
 
-			logger.Debugf("pushsync: could not push to peer %s: %v", peer, err)
+			ps.metrics.TotalSendAttempts.Inc()
 
-			// if the node has warmed up AND no other closer peer has been tried
-			if ps.warmedUp() && timeToSkip > 0 {
-				ps.skipList.Add(ch.Address(), peer, timeToSkip)
-				ps.metrics.TotalSkippedPeers.Inc()
-				logger.Debugf("pushsync: adding to skiplist peer %s", peer.String())
+			skipPeers = append(skipPeers, peer)
+
+			ctxd, cancel := context.WithCancel(ctx)
+			// cancel only after defaultTTL to allow pushPeer to fully complete for inflight requests
+			time.AfterFunc(defaultTTL, cancel)
+
+			go ps.pushPeer(ctxd, resultChan, peer, ch, origin)
+
+			if retries <= 0 {
+				return nil, ErrNoPush
 			}
+
+			// retry
+			timer.Reset(p90TTL)
```

Comment:
why not a simple ticker?

Reply:
If an attempt fails under 5 seconds, we shouldn't have to wait till the next tick.
With a timer its easier to control when the next request should fire.

Reply:
makes sense :+1: